external-crates/move/move-model/src/model.rs

```diff
@@ -3735,6 +3821,123 @@ impl<'env> FunctionEnv<'env> {
+        let mut transferred = BTreeSet::new();
+        let module = &self.module_env.data.module;
+        for b in self.get_bytecode() {
+            if let Bytecode::CallGeneric(fi_idx) = b {
+                let FunctionInstantiation {
+                    handle,
+                    type_parameters,
+                } = module.function_instantiation_at(*fi_idx);
+                let f_ref = FunctionRef::from_idx(module, handle);
+                if is_framework_function(&f_ref, "transfer", vec!["transfer", "public_transfer"]) {
+                    let type_params = module.signature_at(*type_parameters);
+                    transferred.insert(self.module_env.globalize_signature(&type_params.0[0]));
+                }
+            }
+        }
+        if transitive {
+            let callees = self.get_transitive_closure_of_called_functions();
+            for callee in callees {
+                let fenv = self.module_env.env.get_function(callee);
+                transferred.extend(fenv.get_shared_objects(false));
+            }
+        }
+        transferred
+    }
+
+    /// Returns the object types that may be frozen by this function
+    /// If `transitive` is false, only return objects directly frozen by this function
+    /// If `transitive` is true, return objects frozen by both this function and its transitive callees
+    pub fn get_frozen_objects(&'env self, transitive: bool) -> BTreeSet<Type> {
+        let module = &self.module_env.data.module;
+        let mut frozen = BTreeSet::new();
+        for b in self.get_bytecode() {
+            if let Bytecode::CallGeneric(fi_idx) = b {
+                let FunctionInstantiation {
+                    handle,
+                    type_parameters,
+                } = module.function_instantiation_at(*fi_idx);
+                let f_ref = FunctionRef::from_idx(module, handle);
+                if is_framework_function(
+                    &f_ref,
+                    "transfer",
+                    vec!["freeze_object", "public_freeze_object"],
+                ) {
+                    let type_params = module.signature_at(*type_parameters);
+                    frozen.insert(self.module_env.globalize_signature(&type_params.0[0]));
+                }
+            }
+        }
+        if transitive {
+            let callees = self.get_transitive_closure_of_called_functions();
+            for callee in callees {
+                let fenv = self.module_env.env.get_function(callee);
+                frozen.extend(fenv.get_shared_objects(false));
+            }
+        }
+        frozen
+    }
+
+    /// Returns the event types that may be emitted by this function
+    /// If `transitive` is false, only return events directly emitted by this function
+    /// If `transitive` is true, return events emitted by both this function and its transitive callees
+    pub fn get_events(&'env self) -> BTreeSet<Type> {
```

Comment:
looks like this function is missing a `transitive` parameter and logic 😄 

(or maybe `get_events` should only work for direct functions?)

Reply:
Good point--I want transitive here too.